File: firefly/src/main/java/com/firefly/server/http2/router/impl/ParameterPathMatcher.java
package com.firefly.server.http2.router.impl;

import com.firefly.server.http2.router.Matcher;
import com.firefly.server.http2.router.Router;
import com.firefly.server.http2.router.utils.PathUtils;

import java.util.*;

/**
 * @author Pengtao Qiu
 */
public class ParameterPathMatcher implements Matcher {

    private Map<Integer, Map<ParameterPath, Set<Router>>> parameterPath;

    private static class ParameterPath {
        final String rule;
        final List<String> paths;

        public ParameterPath(String rule) {
            this.rule = rule;
            paths = PathUtils.split(rule);
        }

        @Override
        public boolean equals(Object o) {
            if (this == o) return true;
            if (o == null || getClass() != o.getClass()) return false;
            ParameterPath that = (ParameterPath) o;
            return Objects.equals(rule, that.rule);
        }

        @Override
        public int hashCode() {
            return Objects.hash(rule);
        }

        Map<String, String> match(List<String> list) {
            Map<String, String> param = new HashMap<>();
            for (int i = 0; i < list.size(); i++) {
                String path = paths.get(i);
                String value = list.get(i);

                if (path.charAt(0) != ':') {
                    if (!path.equals(value)) {
                        return null;
                    }
                } else {
                    param.put(path.substring(1), value);
                }
            }
            return param;
        }
    }

    private Map<Integer, Map<ParameterPath, Set<Router>>> parameterPath() {
        if (parameterPath == null) {
            parameterPath = new HashMap<>();
        }
        return parameterPath;
    }

    @Override
    public void add(String rule, Router router) {
        ParameterPath parameterPath = new ParameterPath(rule);
        parameterPath().computeIfAbsent(parameterPath.paths.size(), k -> new HashMap<>())
                       .computeIfAbsent(parameterPath, k -> new HashSet<>())
                       .add(router);
    }

    @Override
    public MatchResult match(String value) {
        if (parameterPath == null) {
            return null;
        }

        if (value.length() == 1) {
            if (value.charAt(0) == '/') {
                return null;
            } else {
                throw new IllegalArgumentException("the path: [" + value + "] format error");
            }
        } else {
            List<String> list = PathUtils.split(value);
            Map<ParameterPath, Set<Router>> map = parameterPath.get(list.size());
            if (map != null && !map.isEmpty()) {
                Set<Router> routers = new HashSet<>();
                Map<Router, Map<String, String>> parameters = new HashMap<>();

                map.forEach((key, regRouter) -> {
                    Map<String, String> param = key.match(list);
                    if (param != null) {
                        routers.addAll(regRouter);
                        regRouter.forEach(router -> parameters.put(router, param));
                    }
                });

                if (!routers.isEmpty()) {
                    return new MatchResult(routers, parameters, getMatchType());
                } else {
                    return null;
                }
            } else {
                return null;
            }
        }
    }

    @Override
    public MatchType getMatchType() {
        return MatchType.PATH;
    }
}

File: firefly-net/src/main/kotlin/com/fireflysource/net/http/server/impl/matcher/ParameterPathMatcher.kt
package com.fireflysource.net.http.server.impl.matcher

import com.fireflysource.net.http.server.Matcher
import com.fireflysource.net.http.server.Router
import java.util.*


class ParameterPathMatcher : AbstractMatcher<ParameterPathMatcher.ParameterRule>(), Matcher {

    companion object {
        fun isParameterPath(path: String): Boolean {
            val paths = split(path)
            return paths.any { it[0] == ':' }
        }

        fun split(path: String): List<String> {
            val paths: MutableList<String> = LinkedList()
            var start = 1
            val last = path.lastIndex

            for (i in 1..last) {
                if (path[i] == '/') {
                    paths.add(path.substring(start, i).trim())
                    start = i + 1
                }
            }

            if (path[last] != '/') {
                paths.add(path.substring(start).trim())
            }
            return paths
        }
    }

    inner class ParameterRule(val rule: String) {

        val paths = split(rule)

        fun match(list: List<String>): Map<String, String> {
            if (paths.size != list.size) return emptyMap()

            val param: MutableMap<String, String> = HashMap()
            for (i in list.indices) {
                val path = paths[i]
                val value = list[i]
                if (path[0] != ':') {
                    if (path != value) {
                        return emptyMap()
                    }
                } else {
                    param[path.substring(1)] = value
                }
            }
            return param
        }

        override fun equals(other: Any?): Boolean {
            if (this === other) return true
            if (javaClass != other?.javaClass) return false

            other as ParameterRule
            return rule == other.rule
        }

        override fun hashCode(): Int {
            return rule.hashCode()
        }
    }

    override fun getMatchType(): Matcher.MatchType {
        return Matcher.MatchType.PATH
    }

    override fun add(rule: String, router: Router) {
        val parameterRule = ParameterRule(rule)
        routersMap.computeIfAbsent(parameterRule) { TreeSet() }.add(router)
    }

    override fun match(value: String): Matcher.MatchResult? {
        if (routersMap.isEmpty()) return null

        val routers = TreeSet<Router>()
        val parameters = HashMap<Router, Map<String, String>>()
        val paths = split(value)

        routersMap.forEach { (rule, routerSet) ->
            val param = rule.match(paths)
            if (param.isNotEmpty()) {
                routers.addAll(routerSet)
                routerSet.forEach { router -> parameters[router] = param }
            }
        }
        return if (routers.isEmpty()) null else Matcher.MatchResult(routers, parameters, matchType)
    }
}